pkg/pulumiyaml/syntax/diags.go

```diff
@@ -121,3 +125,16 @@ func (d Diagnostics) Unshown() *Diagnostics {
 	}
 	return &diags
 }
+
+// Inform the user that they did not conform with the expected capitalization style. If
+// `expected` matches `found`, then `nil` is returned. This allows
+// `Diagnostics.Extend(UnexpectedCasing(location, expected, found))` without checking if
+// expected equals found.
+func UnexpectedCasing(rng *hcl.Range, expected, found string) *Diagnostic {
+	if expected == found {
+		return nil
+	}
+	summary := fmt.Sprintf("'%s' looks like a miscapitalization of '%s'", found, expected)
+	detail := "Pulumi YAML will enforce camelCase capitalization by GA. See https://github.com/pulumi/pulumi-yaml/issues/355 for details."
```

Comment:
We may punt on enforcing this by GA, to allow more time for users to fix programs.

```suggestion
	detail := "A future version of Pulumi YAML will enforce camelCase fields. See https://github.com/pulumi/pulumi-yaml/issues/355 for details."
```